src/virtual-machine.js

```diff
@@ -543,17 +543,57 @@ class VirtualMachine extends EventEmitter {
     /**
      * Get an SVG string from storage.
      * @param {int} costumeIndex - the index of the costume to be got.
-     * @return {string} the costume's SVG string, or null if it's not an SVG costume.
+     * @return {string|HTMLImageElement} the costume's SVG string if it's SVG,
+     *     an HTMLImageElement if it's a PNG, or null if it couldn't be found or decoded.
      */
     getCostume (costumeIndex) {
         const id = this.editingTarget.getCostumes()[costumeIndex].assetId;
-        if (id && this.runtime && this.runtime.storage &&
-                this.runtime.storage.get(id).dataFormat === 'svg') {
+        if (!id || !this.runtime || !this.runtime.storage) return null;
+        const format = this.runtime.storage.get(id).dataFormat;
+        if (format === this.runtime.storage.DataFormat.SVG) {
             return this.runtime.storage.get(id).decodeText();
+        } else if (format === this.runtime.storage.DataFormat.PNG ||
+                format === this.runtime.storage.DataFormat.JPG) {
+            const data = this.runtime.storage.get(id).encodeDataURI();
+            const image = new Image();
+            image.src = data;
+            return image;
```

Comment:
Technically speaking, the image might not be ready by the time this function returns. In order to guarantee compatibility and avoid race conditions, I would recommend adjusting `getCostume` to return a `Promise` which you'd `resolve` with the return value in the image's `onload` callback. This would also give you a chance to `reject` the promise in the image's `onerror` callback. That'll probably make callers more complex, but I don't think the current code is guaranteed to be safe.

Reply:
Which one of these calls is async? encodeDataURI?

Reply:
That would be fine, but it also seems like just returning the encoded data URI, which is sync, would be fine too. It would also more closely parallel the SVG return (also string, also needs to be applied to an image tag to view, etc.)

Reply:
`image.src = data;` is async in that, in general, it may make a network request before becoming populated. I'm not aware of a special-case guarantee that a data URI will be decoded synchronously, but if that's guaranteed then there's no problem here. I assume this code has been tested successfully at least once, which suggests that I might be missing something ;)

Other possibilities: it might be that decoding a data URI is unofficially synchronous on some browsers, or it might just be quick enough for smallish images on faster computers.

Reply:
@cwillisf that makes sense, had a similar conversation with @paulkaplan offline as well, thank you both for explaining =]